core/src/main/java/com/graphhopper/GraphHopper.java

```diff
@@ -1238,7 +1248,90 @@ protected void postProcessing(boolean closeEarly) {
             loadOrPrepareCH(closeEarly);
     }
 
+
     protected void importPublicTransit() {
+
+        //We use that method to import speeds
+        if(speedsProvider != null){
+            logger.info("Start Custom Speeds Provider");
+            long startTime = System.nanoTime();
+
+            List<String> vehicles = this.getProfiles().stream().map(Profile::getVehicle).distinct().collect(Collectors.toList());
+
+            List<DecimalEncodedValue> speedEncoders = vehicles.stream().map(
+                    vehicle -> this.getEncodingManager().getDecimalEncodedValue(VehicleSpeed.key(vehicle))).collect(Collectors.toList());
+            setSpeedsFor(speedEncoders);
+
+            long endTime = System.nanoTime();
+            long elapsed = (endTime - startTime);
+            long durationInMillis = TimeUnit.NANOSECONDS.toMillis(elapsed);
+            logger.info("End Custom Speeds Provider : Graph processed in " + durationInMillis);
+        }
+    }
+
+    private Optional<Double> speedFor(int osmWayId, RoadClass roadClass) {
+
+        Optional<Double>  waySpeed = speedsProvider.speedForWay(osmWayId);
+
+        if(waySpeed.isPresent()){
+            return waySpeed;
+        }else {
+            return speedsProvider.speedForRoadClass(roadClass);
+        }
+    }
+
+    private void setSpeedsFor(List<DecimalEncodedValue> speedEncoders) {
+        IntEncodedValue OSMWayIDEncoder = this.getEncodingManager().getIntEncodedValue(OSMWayID.KEY);
+        DecimalEncodedValue maxSpeedEncoder = this.getEncodingManager().getDecimalEncodedValue(MaxSpeed.KEY);
+        EnumEncodedValue<RoadClass> roadClassEncoder = this.getEncodingManager().getEnumEncodedValue(RoadClass.KEY,RoadClass.class);
+
+        AllEdgesIterator iter = this.getBaseGraph().getAllEdges();
+        while(iter.next()){
+
+            int edge = iter.getEdge();
+
+            //Normal Direction
+            double maxSpeed = iter.get(maxSpeedEncoder);
+            int osmWayId = iter.get(OSMWayIDEncoder);
+            RoadClass roadClass = iter.get(roadClassEncoder);
+            Optional<Double> maybeCustomSpeed = speedFor(osmWayId,roadClass);
```

Comment:
I have doubts about the Error Handling. Is it possible the SpeedsProvider launch unchecked exceptions right? As it is connected to an external storage s3...so if we don't want to fail in front of that unchecked exceptions, we should capture something here no?

Reply:
Speeds Provider return an Option, in theory it will manage it the exceptions and in case of any unexpected error return a None, WDYT?